### PR TITLE
add support for workload identity

### LINKED ISSUE #2205: Migrate from AAD pod identity to Azure Workload Identity
/kind feature

[Azure AD Workload Identity](https://github.com/Azure/azure-workload-identity) is the next iteration of [Azure AD Pod Identity](https://github.com/Azure/aad-pod-identity) that enables Kubernetes applications to access Azure cloud resources securely with [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/) based on annotated [service accounts](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/). 

Since Azure AD Pod Identity is getting deprecated, we should migrate all CAPZ usage to Azure Workload Identity.

### PR DESCRIPTION
**Pull request by sonasingh46**
**What type of PR is this?**
This PR enables capz to use Azure Workload Identity. 

**What this PR does / why we need it**:
Implementation of proposal https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2814

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #2205 

**Special notes for your reviewer**:
- This PR should be merged after merging https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2748[Done]
- The test Jobs should pass once preset is enabled on all the Jobs. [Done]
- We will need to convert all the templates to using user-assigned-identity. [Done]
- Move all the e2e tests to be using workload identity. [Done] 
- Current developer experience changes with the current change for tilt. `make tilt-up`  with the current changes will require all the devs to set up the workload identity pre-requisites and expose 3 env variables. Do you think we should continue with this or have some env vars e.g. `ENABLE_WORKLOAD_IDENTITY=true` explicitly specified to do so. And if not, the usual way of make tilt-up will work. [Needs Review/Discussion]

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Enhancement to use Azure Workload Identity 
```

### CODE REVIEW

**Comment by sonasingh46** [comment]
@CecileRobertMichon -- I have worked on top of this PR https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2748 commit of @r4f4 

Once that PR gets merged, I will rebase it. 

Pending Items to remove WIP : 
- CI setup for azwi which requires storing key pair and publishing OIDC URL
- A way to copy key pair on the workload cluster as IIRC one e2e test creates a workload cluster and then converts it for management cluster for v1alpha3 --> v1alpha4 upgrade. 

Feedback, suggestion are  welcome. 

cc @jackfrancis @mboersma @aramase

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -509,6 +509,9 @@ const (
 
 	// ServicePrincipalCertificate represents a service principal using a certificate as secret.
 	ServicePrincipalCertificate IdentityType = "ServicePrincipalCertificate"
+
+	// WorkloadIdentity represents a WorkloadIdentity.
+	WorkloadIdentity IdentityType = "WorkloadIdentity"
```
Comment (CecileRobertMichon) [inline api/v1beta1/types.go:514]: wonder if this is an appropriate name for this identity "type" or if the name should be based more on the Azure identity resource (`FederatedIdentityCredential`) as opposed to the k8s project that enables this (`WorkloadIdentity`) as it's more of an implementation detail.

Also do we foresee Workload Identity supporting more than one "type" in the future? For example, it looks like you could already use either AD application or User Assigned Identity https://azure.github.io/azure-workload-identity/docs/topics/federated-identity-credential.html#federated-identity-credential-for-a-user-assigned-managed-identity, so we should be able to differentiate between those (unless the code to use them in CAPZ is the same)? @aramase thoughts?

Comment (aramase) [reply api/v1beta1/types.go:514]: > Also do we foresee Workload Identity supporting more than one "type" in the future? For example, it looks like you could already use either AD application or User Assigned Identity https://azure.github.io/azure-workload-identity/docs/topics/federated-identity-credential.html#federated-identity-credential-for-a-user-assigned-managed-identity, so we should be able to differentiate between those (unless the code to use them in CAPZ is the same)? @aramase thoughts?

The code to use workload identity federation would be the same irrespective of the underlying Azure AD object (application or managed identity) it's tied to. In terms of the name, I think [Workload Identity Federation](https://learn.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation) is used commonly.

Comment (sonasingh46) [reply api/v1beta1/types.go:514]: Sure. Like @aramase said -- it will really not matter if we use an AD application or User Assigned Identity as far as CAPZ implementation is concerned. 
Also I consider `Workload Identity Federation` as one of the entities/workflows of the entire `Workload Identity` thing. 
And `Workload Identity` will be there for other providers too ( if not yet there ) though the implementations and terminology under the hood may differ. 

So `WorkloadIdentity` looks simple, and intuitive to me from a user's POV. 

Just to take decision -- we are deciding b/w `WorkloadIdentity` vs `WorkloadIdentityFederation` , right?

File: scripts/kind-with-registry.sh [hunk 2]
```diff
@@ -40,10 +40,38 @@ if [ "${running}" != 'true' ]; then
   docker run -d --restart=always -p "127.0.0.1:${reg_port}:5000" --name "${reg_name}" registry:2
 fi
 
+# Set up the key pair file paths for azwi in CI
+# SERVICE_ACCOUNT_SIGNING_PUB and SERVICE_ACCOUNT_SIGNING_KEY are preset envs
+# Refer: https://github.com/kubernetes/test-infra/pull/28303/files
+
+pub=$(echo ${SERVICE_ACCOUNT_SIGNING_PUB} | base64 --decode)
```
Comment (sonasingh46) [inline scripts/kind-with-registry.sh:47]: @CecileRobertMichon -- Getting this error here. 
```
./scripts/kind-with-registry.sh: line 47: SERVICE_ACCOUNT_SIGNING_PUB: unbound variable
```
How can I convert what is available in in the env to a file so that kind cluster gets to mount this file. Also is this env accessible for the process launched by this script?

File: config/rbac/service_account.yaml [hunk 3]
```diff
@@ -1,5 +1,7 @@
 apiVersion: v1
 kind: ServiceAccount
 metadata:
+  labels:
+    azure.workload.identity/use: "true"
```
Comment (sonasingh46) [inline config/rbac/service_account.yaml:5]: Note to self: Move this to pod. Annotation on SA no longer required.

File: scripts/kind-with-registry.sh [hunk 4]
```diff
@@ -40,10 +40,38 @@ if [ "${running}" != 'true' ]; then
...
+    - hostPath: $HOME/azwi/creds/sa.pub
+      containerPath: /etc/kubernetes/pki/sa.pub
+    - hostPath: $HOME/azwi/creds/sa.key
+      containerPath: /etc/kubernetes/pki/sa.key
+  kubeadmConfigPatches:
+  - |
+    kind: ClusterConfiguration
+    apiServer:
+      extraArgs:
+        service-account-issuer: ${SERVICE_ACCOUNT_ISSUER}
```
Comment (sonasingh46) [inline scripts/kind-with-registry.sh:69]: @CecileRobertMichon -- Will need a preset for this one too.

File: scripts/kind-with-registry.sh (hunk 2 above)
Comment (sonasingh46) [reply scripts/kind-with-registry.sh]: Oh.. Never mind. I got confused. So the kind(bootstrap cluster ) for e2e test are created via the capi library and looks like there is not option to pass the flags for kind cluster that is needed.

**Comment by sonasingh46** [comment]
@CecileRobertMichon @mboersma -- Shall we go with any workaround for passing required configuration for kind cluster. ( One way is to have the code for this locally in capz -- but this still looks more work ) or just raise a PR in capi to help enable that. 

PS: For workload identity, I need to be able to set up extraMounts and put kubeadmconfigpatches on bootstrap cluster and looks like the current e2e framework does not have the way that allows this.

**Comment by sonasingh46** [comment]
Also, to deploy azure workload identity on bootstrap cluster -- the current e2e code has again dependency with the capi e2e framework. One way again here is to have our own way of deploying from capz. 
Would love to get some guidance on how to proceed on these.

**Comment by dlipovetsky** [comment]
@sonasingh46 Thanks for working on this!

In addition to the "create a cluster with workload identity" e2e test, I think it would be very useful to have a test to verify that CAPZ only uses the identity referenced by a cluster.

For example, a test that
- Creates two namespaces
- Creates an AzureClusterIdentity an identical in each namespace
- Creates a Secret with a working client secret in one namespace, and a Secret with a bad (i.e. not working) client secret in the other namespace.
- Create clusters in both namespaces
- Waits for one cluster to be created, and some more time to confirm that the other cluster is _not_ created.

I'm happy to follow up with more details, and even contribute this kind of test in the future.

File: Makefile [hunk 5]
```diff
@@ -678,6 +678,10 @@ test-e2e: ## Run "docker-build" and "docker-push" rules then run e2e tests.
 	$(MAKE) docker-build docker-push \
 	test-e2e-run
 
+create-bootstrap-cluster:
+	KIND_CLUSTER_NAME=capz-e2e && ./scripts/kind-with-registry.sh
+
+
```
Comment (aramase) [inline Makefile:684]:
```suggestion
```

File: Tiltfile [hunk 6]
```diff
@@ -21,6 +21,7 @@ settings = {
     "kind_cluster_name": "capz",
     "capi_version": "v1.4.1",
     "cert_manager_version": "v1.11.0",
+    "azwi_version": "v0.14.0",
```
Comment (aramase) [inline Tiltfile:24]: The latest release is `v1.0.0`.

File: azure/scope/workload_identity.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (aramase) [inline azure/scope/workload_identity.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: azure/scope/workload_identity.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/pkg/errors"
+)
+
+/*
+
+AZWI (Azure Workload Identity) required deploying AZWI mutating admission webhook
```
Comment (aramase) [inline azure/scope/workload_identity.go:33]:
```suggestion
Azure Workload Identity (AZWI) requires deploying AZWI mutating admission webhook
```

File: azure/scope/workload_identity.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+for self managed clusters e.g. Kind.
+
+The webhook injects the following environment variables to the pod that
+uses a service account with a annotation `azure.workload.identity/use=true`
```
Comment (aramase) [inline azure/scope/workload_identity.go:37]: The comment needs to be updated!

The behavior was changed in `v0.15.0` and later. **Only** pods with label `azure.workload.identity/use=true` will be mutated (with environment variables and projected service account volume). The label/annotation is not required in the service account.

File: azure/scope/workload_identity.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+|-----------------------------------------------------------------------------------|
+|AZURE_AUTHORITY_HOST       | The Azure Active Directory (AAD) endpoint.            |
+|AZURE_CLIENT_ID            | The application/client ID of the Azure AD             |
```
Comment (aramase) [inline azure/scope/workload_identity.go:40]:
```suggestion
|AZURE_CLIENT_ID            | The client ID of the Azure AD             |
```

File: azure/scope/workload_identity.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+In addition to the service account, it also projects a signed service account token to the
+workload's volume(in this case the capz pod). The volume name is `azure-identity-token`
+which is mounted at path `/var/run/secrets/azure/tokens/azure-identity-token` to the pod.
+
+*/
+
+const (
+	// AzureFedratedTokenFileENVKey is the env key for AZURE_FEDERATED_TOKEN_FILE.
+	AzureFedratedTokenFileENVKey = "AZURE_FEDERATED_TOKEN_FILE"
```
Comment (aramase) [inline azure/scope/workload_identity.go:54]: nit: `s/AzureFedratedTokenFileENVKey/AzureFedratedTokenFileEnvKey`

similar comment for the other constants.

File: azure/scope/workload_identity.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+// WorkloadIdentityCredentialOptions contains the configurable options for azwi.
+type WorkloadIdentityCredentialOptions struct {
+	ClientID      string
+	TenantID      string
+	TokenFilePath string
+	azcore.ClientOptions
```
Comment (aramase) [inline azure/scope/workload_identity.go:73]: nit: move the embedding to the first line in the struct

File: azure/scope/workload_identity.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+// WithDefaults sets token file path. It also sets the client tenant ID from injected env in
+// case empty values are passed.
+func (w *WorkloadIdentityCredentialOptions) WithDefaults() (*WorkloadIdentityCredentialOptions, error) {
+	tokenFilePath, err := GetProjectedTokenPath()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get token file path for identity")
+	}
+	w.TokenFilePath = tokenFilePath
```
Comment (aramase) [inline azure/scope/workload_identity.go:109]: Need to check if the token file path is not empty here?

File: azure/scope/workload_identity.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+	// Fallback to using client ID from env variable if not set.
+	if strings.TrimSpace(w.ClientID) == "" {
+		w.ClientID = os.Getenv(AzureClientIDENVKey)
```
Comment (aramase) [inline azure/scope/workload_identity.go:113]: Is CAPZ going to always have the user provide the clientID as part of the custom resource spec? If that's the case when will this value ever be not empty?

File: azure/scope/workload_identity.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+		w.TenantID = os.Getenv(AzureTenantIDENVKey)
+		if strings.TrimSpace(w.TenantID) == "" {
+			return nil, errors.New("empty tenant ID")
+		}
+	}
+	return w, nil
+}
+
+// NewWorkloadIdentityCredential returns a workload identity credential.
+func NewWorkloadIdentityCredential(options *WorkloadIdentityCredentialOptions) (*workloadIdentityCredential, error) {
```
Comment (aramase) [inline azure/scope/workload_identity.go:130]: This should return the interface `azcore.TokenCredential` instead of the struct

File: config/default/azwi.yaml [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+      - ""
+    resources:
+      - serviceaccounts
+    verbs:
+      - create
+      - get
+      - list
+      - patch
+      - update
+      - watch
```
Comment (aramase) [inline config/default/azwi.yaml:49]: These permissions have been updated (`create`, `patch` and `update` are not required)

xref: https://github.com/Azure/azure-workload-identity/blob/release-1.0/deploy/azure-wi-webhook.yaml

File: config/default/azwi.yaml [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+          env:
+            - name: POD_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  apiVersion: v1
+                  fieldPath: metadata.namespace
+          envFrom:
+            - configMapRef:
+                name: azure-wi-webhook-config
+          image: mcr.microsoft.com/oss/azure/workload-identity/webhook:v0.15.0
```
Comment (aramase) [inline config/default/azwi.yaml:158]: There have been changes to the yaml (`Role`, `ClusterRole` and some more). I would recommend updating this to the `v1.0.0` deploy yaml.

File: config/manager/manager.yaml [hunk 18]
```diff
@@ -16,6 +16,7 @@ spec:
       labels:
         control-plane: capz-controller-manager
         aadpodidbinding: capz-controller-aadpodidentity-selector
```
Comment (aramase) [inline config/manager/manager.yaml:18]: If there is no plan to support pod identity anymore after workload identity is supported, maybe add a `TODO` or an issue to track removing this label.

File: config/rbac/service_account.yaml (hunk 3 above)
Comment (aramase) [inline config/rbac/service_account.yaml:5]: This can be removed already as you're using `v0.15.0` and hopefully updating to `v1.0.0` as part of this PR

File: test/e2e/config/azwi.yaml [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+          env:
+            - name: POD_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  apiVersion: v1
+                  fieldPath: metadata.namespace
+          envFrom:
+            - configMapRef:
+                name: azure-wi-webhook-config
+          image: mcr.microsoft.com/oss/azure/workload-identity/webhook:v1.0.0-beta.0
```
Comment (aramase) [inline test/e2e/config/azwi.yaml:171]: ditto comment (update the yaml here to the `v1.0.0` release)

**Review by aramase — CHANGES_REQUESTED** [review]

File: Makefile [hunk 20]
```diff
@@ -166,7 +166,7 @@ E2E_CONF_FILE ?= $(ROOT_DIR)/test/e2e/config/azure-dev.yaml
 E2E_CONF_FILE_ENVSUBST := $(ROOT_DIR)/test/e2e/config/azure-dev-envsubst.yaml
 SKIP_CLEANUP ?= false
 SKIP_LOG_COLLECTION ?= false
-SKIP_CREATE_MGMT_CLUSTER ?= false
+SKIP_CREATE_MGMT_CLUSTER ?= true
```
Comment (CecileRobertMichon) [inline Makefile:169]: why this change?

File: Makefile [hunk 21]
```diff
@@ -678,6 +678,10 @@ test-e2e: ## Run "docker-build" and "docker-push" rules then run e2e tests.
 	$(MAKE) docker-build docker-push \
 	test-e2e-run
 
+create-bootstrap-cluster:
```
Comment (CecileRobertMichon) [inline Makefile:681]: how is this different from the `kind-create` target?

File: test/e2e/azure_test.go [hunk 22]
```diff
@@ -903,4 +903,55 @@ var _ = Describe("Workload cluster creation", func() {
 			By("PASSED!")
 		})
 	})
+
+	// Workload identity test
+	Context("Creating a cluster that uses workload identity [OPTIONAL]", func() {
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:908]: Instead of adding a new test for this, could we switch all the existing tests to use Workload Identity? If we want to keep coverage for Legacy ADD Pod Identity for now we can add an optional test for it.

File: test/e2e/config/azwi.yaml [hunk 23]
```diff
@@ -0,0 +1,272 @@
+apiVersion: v1
```
Comment (CecileRobertMichon) [inline test/e2e/config/azwi.yaml:1]: why is this file needed?

File: test/e2e/config/azwi.yaml [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+  name: azure-wi-webhook-manager-role
+subjects:
+  - kind: ServiceAccount
+    name: azure-wi-webhook-admin
+    namespace: azure-workload-identity-system
+---
+apiVersion: v1
+data:
+  AZURE_ENVIRONMENT: AzurePublicCloud
+  AZURE_TENANT_ID: b39138ca-3cee-4b4a-a4d6-cd83d9dd62f0
```
Comment (CecileRobertMichon) [inline test/e2e/config/azwi.yaml:110]: why is there data for TENANT_ID checked in here?

File: test/e2e/e2e_suite_test.go [hunk 25]
```diff
@@ -69,6 +69,8 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	clusterctlConfigPath = createClusterctlLocalRepository(e2eConfig, filepath.Join(artifactFolder, "repository"))
 
 	By("Setting up the bootstrap cluster")
+	// ToDo
```
Comment (CecileRobertMichon) [inline test/e2e/e2e_suite_test.go:72]: is this still TODO?

File: test/e2e/e2e_suite_test.go [hunk 26]
```diff
@@ -163,15 +165,26 @@ func setupBootstrapCluster(config *clusterctl.E2EConfig, useExistingCluster bool
...
 	clusterProxy := NewAzureClusterProxy("bootstrap", kubeconfigPath)
 	Expect(clusterProxy).NotTo(BeNil(), "Failed to get a bootstrap cluster proxy")
-
 	return clusterProvider, clusterProxy
 }
 
 func initBootstrapCluster(bootstrapClusterProxy framework.ClusterProxy, config *clusterctl.E2EConfig, clusterctlConfig, artifactFolder string) {
+	// This deploys azwi webhook. It is important that azwi webhook config gets deployed before capz config as capz
+	// depends on env var and projected service account token volume.
+	// ToDo: @sonasingh46 ; find some better way to do it.
```
Comment (CecileRobertMichon) [inline test/e2e/e2e_suite_test.go:186]: is this still TODO?

**Comment by dlom** [comment]
@sonasingh46 need any help with getting this PR moving along? We are interested in upstreaming it once this change merges

File: test/e2e/azure_test.go (hunk 22 above)
Comment (sonasingh46) [reply test/e2e/azure_test.go:908]: I think the initial plan was to move the functionality to working state and then switch completely in the next PR. Advantage is, this PR is smaller and gets some soak time. 
So we will need to move to using user assigned templates and move to workload identity in the next PR so that it is easy to triage/debug failures if any.
Step by step is what my through process is here. wdyt?

File: test/e2e/config/azwi.yaml (hunk 24 above)
Comment (sonasingh46) [reply test/e2e/config/azwi.yaml:110]: That is a miss. Will remove that.

File: Makefile (hunk 20 above)
Comment (sonasingh46) [reply Makefile:169]: Keeping this var to `true` for workload identity. As for workload identity we need to create a kind cluster
by passing explicit flags for key pairs ( see line 75-80 in ./scripts/kind-with-registry.sh )
Once the e2e framework supports this we can move back to setting it `false`

Will add this as a comment too.

File: Makefile (hunk 21 above)
Comment (sonasingh46) [reply Makefile:681]: Almost similar -- just a bit of difference is that `kind-create` is used by tilt and in this target we explicitly set the cluster name to be capz-e2e. 
I kept those different to kind of segregate the local dev and e2e scenarios.

File: test/e2e/config/azwi.yaml (hunk 23 above)
Comment (sonasingh46) [reply test/e2e/config/azwi.yaml:3]: This config is used to deploy azwi webhook in e2e 
Updated the PR to include a comment also.

File: azure/scope/workload_identity.go (hunk 13 above)
Comment (sonasingh46) [reply azure/scope/workload_identity.go:109]: GetProjectedTokenPath function above at line 105 always returns a path or error. it can't be empty.

**Comment by sonasingh46** [comment]
Added a different PR https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3583 just to enable workload identity capability in capz to reduce the scope of PR. 
Other template changes and moving all ci test to worklaod identity can be done in subsequent PRs. 

Did not change in this PR itself to keep the change ready so that it is easy to pick the change from here for the subsequent PRs.
